src/fs_store.rs

```diff
@@ -0,0 +1,476 @@
+            {
+                tracing::warn!(
+                    non_dir_path=%entry.path().display(),
+                    "unexpected non-directory at level1 of database"
+                );
+                continue;
+            }
+            let level1 = entry.path();
+            let entries = level1
+                .read_dir()
+                .map_err(|e| Error(ErrorKind::ErrReadingLevel1Path(level1.clone(), e)))?;
+            for entry in entries {
+                let entry =
+                    entry.map_err(|e| Error(ErrorKind::ErrReadingLevel1Path(level1.clone(), e)))?;
+                let metadata = entry
+                    .metadata()
+                    .map_err(|e| Error(ErrorKind::ErrReadingLevel2Path(entry.path(), e)))?;
+                if metadata.is_dir() {
+                    tracing::warn!(
+                        non_file_path=%entry.path().display(),
+                        "unexpected directory at level2 of database"
+                    );
+                    continue;
+                }
+                let Some(doc_paths) = DocIdPaths::parse(&level1, entry.path()) else {
+                    tracing::warn!(
+                        non_doc_path=%entry.path().display(),
+                        "unexpected non-document path at level2 of database"
+                    );
+                    continue;
+                };
+                result.insert(doc_paths.doc_id);
+            }
+        }
+        Ok(result.into_iter().collect())
+    }
+
+    pub fn append(&self, id: &DocumentId, changes: &[u8]) -> Result<(), Error> {
+        let paths = DocIdPaths::from(id);
+        std::fs::create_dir_all(paths.level2_path(&self.root)).map_err(|e| {
+            Error(ErrorKind::CreateLevel2Path(
+                paths.level2_path(&self.root),
+                e,
+            ))
+        })?;
+
+        let chunk_name = SavedChunkName::new_incremental(changes);
+        let chunk_path = paths.chunk_path(&self.root, &chunk_name);
+
+        tracing::trace!(path=?chunk_path, "writing chunk to filesystem");
+        std::fs::write(&chunk_path, changes)
+            .map_err(|e| Error(ErrorKind::WriteSnapshot(chunk_path.clone(), e)))?;
+        Ok(())
+    }
+
+    pub fn compact(&self, id: &DocumentId, _full_doc: &[u8]) -> Result<(), Error> {
+        let paths = DocIdPaths::from(id);
+
+        // Load all the data we have into a doc
+        let Some(chunks) = Chunks::load(&self.root, id)? else {
+            tracing::warn!(doc_id=%id, "attempted to compact non-existent document");
+            return Ok(())
+        };
+        let mut doc = chunks
+            .to_doc()
+            .map_err(|e| Error(ErrorKind::LoadDocToCompact(e)))?;
+
+        // wrtie the compacted doc to a temporary file
+        let saved = doc.save();
+        let temp_save =
+            tempfile::NamedTempFile::new().map_err(|e| Error(ErrorKind::CreateTempFile(e)))?;
+        std::fs::write(temp_save.path(), saved)
```

Comment:
In order to ensure no data can be lost, should we not use something like [`sync_all`](https://doc.rust-lang.org/std/fs/struct.File.html#method.sync_all) on this before deleting the older data below?

Reply:
Yes good point! Technically on `ext4` with default mount options this isn't needed due to the `auto_da_alloc` flag, but we really shouldn't be relying on that. Force push incoming.

Reply:
pushed

Reply:
Been thinking about whether we need to do this on append as well. I _think_ we can get away without the fsync though. If we write appends to a tempfile and then rename, but _don't_ do the fsync then if the system crashes we could end up with a zero length incremental. This would be lost data but the document would still load. I'm inclined to think this kind of data loss is acceptable for incrementals but @issackelly I would be interested in whether that would be suitable for your usecase?

Reply:
Sorry I mean "filesystem sync" not "automerge sync". I am in favor of having the ability for requests to persist to confirm their persistence. If there is a performance hit it could be FastFsStorage / DurableFsStorage instead of configuration paramaters as well, if that's your preference.

Reply:
I've just made this fsync on append and compact. If that causes performance problems we can maybe add a config parameter.